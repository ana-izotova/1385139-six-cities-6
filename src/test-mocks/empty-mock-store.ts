import {AuthorizationStatus, Cities, FetchStatus, SortType, NameSpace} from "../const";
import {UserInitialStateTypes} from "../store/user/user-types";
import {AllOffersInitialStateTypes} from "../store/all-offers/all-offers-types";
import {SingleOfferInitialStateTypes} from "../store/single-offer/single-offer-types";
import {FavoritesInitialStateTypes} from "../store/favorites/favorites-types";

const userEmptyState: UserInitialStateTypes = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  login: ``,
  userAvatar: ``,
  fetchStatus: FetchStatus.INIT
};

const allOffersEmptyState: AllOffersInitialStateTypes = {
  currentCity: Cities[0],
  allOffers: [],
  currentSort: SortType.POPULAR,
  isDataLoaded: false,
  fetchStatus: FetchStatus.INIT,
  favoritesHaveBeenChanged: false
};

const singleOfferEmptyState: SingleOfferInitialStateTypes = {
  offer: null,
  offersNearby: [],
  comments: [],
  isOfferLoaded: false,
  fetchStatus: FetchStatus.INIT
};

const favoritesEmptyState: FavoritesInitialStateTypes = {
  favoriteCards: [],
  areFavoriteCardsLoaded: false
};

export const emptyMockStore = {
  [NameSpace.ALL_OFFERS]: allOffersEmptyState,
  [NameSpace.SINGLE_OFFER]: singleOfferEmptyState,
  [NameSpace.USER]: userEmptyState,
  [NameSpace.FAVORITES]: favoritesEmptyState
};
