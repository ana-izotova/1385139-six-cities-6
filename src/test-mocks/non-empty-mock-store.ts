import {AuthorizationStatus, Cities, FetchStatus, NameSpace, SortType} from "../const";
import {UserInitialStateTypes} from "../store/user/user-types";
import {SingleOfferInitialStateTypes} from "../store/single-offer/single-offer-types";
import {currentOffer, offersNearby, comments, allOffers} from "./adapted-data-mock";
import {AllOffersInitialStateTypes} from "../store/all-offers/all-offers-types";
import {FavoritesInitialStateTypes} from "../store/favorites/favorites-types";

const userStore: UserInitialStateTypes = {
  authorizationStatus: AuthorizationStatus.AUTH,
  login: `jondoe@mail.com`,
  userAvatar: `avatar.jpg`,
  fetchStatus: FetchStatus.DONE
};

const singleOfferStore: SingleOfferInitialStateTypes = {
  offer: currentOffer,
  offersNearby,
  comments,
  isOfferLoaded: true,
  fetchStatus: FetchStatus.DONE
};

const allOffersStore: AllOffersInitialStateTypes = {
  currentCity: Cities[0],
  allOffers,
  currentSort: SortType.POPULAR,
  isDataLoaded: true,
  fetchStatus: FetchStatus.DONE,
  favoritesHaveBeenChanged: true
};

const favoritesStore: FavoritesInitialStateTypes = {
  favoriteCards: [],
  areFavoriteCardsLoaded: false
};

export const nonEmptyMockStore = {
  [NameSpace.ALL_OFFERS]: allOffersStore,
  [NameSpace.SINGLE_OFFER]: singleOfferStore,
  [NameSpace.USER]: userStore,
  [NameSpace.FAVORITES]: favoritesStore
};

