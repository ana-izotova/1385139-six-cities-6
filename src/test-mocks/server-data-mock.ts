export const testOffer = {
  "city": {
    "name": `Paris`,
    "location": {
      "latitude": 48.85661,
      "longitude": 2.351499,
      "zoom": 13,
    },
  },
  "preview_image": `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg`,
  "images": [
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/7.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/8.jpg`,
  ],
  "title": `The house among olive `,
  "is_favorite": false,
  "is_premium": true,
  "rating": 4,
  "type": `room`,
  "bedrooms": 2,
  "max_adults": 2,
  "price": 500,
  "goods": [`Breakfast`, `Air conditioning`],
  "host": {
    "id": 25,
    "name": `Madelina`,
    "is_pro": true,
    "avatar_url": `img/avatar-madelina.jpg`,
  },
  "description": `Relax, rejuvenate and unplug in this ultimate rustic getaway experience in the country.`,
  "location": {
    "latitude": 48.83861,
    "longitude": 2.350499,
    "zoom": 16,
  },
  "id": 1,
};

export const testOffer2 = {
  "city": {
    "name": `Paris`,
    "location": {
      "latitude": 48.85661,
      "longitude": 2.351499,
      "zoom": 13,
    },
  },
  "preview_image": `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg`,
  "images": [
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/7.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/8.jpg`,
  ],
  "title": `The house among olive `,
  "is_favorite": false,
  "is_premium": true,
  "rating": 4,
  "type": `room`,
  "bedrooms": 2,
  "max_adults": 2,
  "price": 500,
  "goods": [`Breakfast`, `Air conditioning`],
  "host": {
    "id": 25,
    "name": `Madelina`,
    "is_pro": true,
    "avatar_url": `img/avatar-madelina.jpg`,
  },
  "description": `Relax, rejuvenate and unplug in this ultimate rustic getaway experience in the country.`,
  "location": {
    "latitude": 48.83861,
    "longitude": 2.350499,
    "zoom": 16,
  },
  "id": 2,
};

export const testComment = {
  "id": 1,
  "user": {
    "id": 15,
    "is_pro": false,
    "name": `Kendall`,
    "avatar_url": `https://assets.htmlacademy.ru/intensives/javascript-3/avatar/6.jpg`,
  },
  "rating": 3,
  "comment": `We loved it so much, the house, the veiw, the location just great.. Highly recommend :)`,
  "date": `2021-02-12T08:04:28.647Z`,
};

export const testOfferFavorited = {
  "city": {
    "name": `Paris`,
    "location": {
      "latitude": 48.85661,
      "longitude": 2.351499,
      "zoom": 13,
    },
  },
  "preview_image": `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg`,
  "images": [
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/7.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/8.jpg`,
  ],
  "title": `The house among olive `,
  "is_favorite": true,
  "is_premium": true,
  "rating": 4,
  "type": `room`,
  "bedrooms": 2,
  "max_adults": 2,
  "price": 500,
  "goods": [`Breakfast`, `Air conditioning`],
  "host": {
    "id": 25,
    "name": `Madelina`,
    "is_pro": true,
    "avatar_url": `img/avatar-madelina.jpg`,
  },
  "description": `Relax, rejuvenate and unplug in this ultimate rustic getaway experience in the country.`,
  "location": {
    "latitude": 48.83861,
    "longitude": 2.350499,
    "zoom": 16,
  },
  "id": 1,
};

export const testOffer2Favorited = {
  "city": {
    "name": `Paris`,
    "location": {
      "latitude": 48.85661,
      "longitude": 2.351499,
      "zoom": 13,
    },
  },
  "preview_image": `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/16.jpg`,
  "images": [
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/7.jpg`,
    `https://assets.htmlacademy.ru/intensives/javascript-3/hotel/8.jpg`,
  ],
  "title": `The house among olive `,
  "is_favorite": true,
  "is_premium": true,
  "rating": 4,
  "type": `room`,
  "bedrooms": 2,
  "max_adults": 2,
  "price": 500,
  "goods": [`Breakfast`, `Air conditioning`],
  "host": {
    "id": 25,
    "name": `Madelina`,
    "is_pro": true,
    "avatar_url": `img/avatar-madelina.jpg`,
  },
  "description": `Relax, rejuvenate and unplug in this ultimate rustic getaway experience in the country.`,
  "location": {
    "latitude": 48.83861,
    "longitude": 2.350499,
    "zoom": 16,
  },
  "id": 2,
};


