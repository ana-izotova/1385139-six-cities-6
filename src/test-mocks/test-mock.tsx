import React from 'react';
import configureStore from 'redux-mock-store';
import {nonEmptyMockStore} from "./non-empty-mock-store";
import * as redux from 'react-redux';
import {Router} from 'react-router-dom';
import {createMemoryHistory} from 'history';
import thunk, {ThunkDispatch} from 'redux-thunk';
import {RootStateType} from "../store/root-reducer";
import {ActionTypes} from "../store/action-types";
import {redirect} from "../middlewares/redirect";
import {emptyMockStore} from "./empty-mock-store";

interface TestProps {
  children?: any,
  pushUrl?: string,
  emptyStore?: boolean
}

export const TestMock: React.FC<TestProps> = ({children, pushUrl, emptyStore = false}) => {
  jest.spyOn(redux, `useDispatch`);
  jest.spyOn(redux, `useSelector`);
  const mockStore = configureStore<RootStateType, ThunkDispatch<RootStateType, any, ActionTypes>>([thunk, redirect]);
  const store = emptyStore ? mockStore(emptyMockStore) : mockStore(nonEmptyMockStore);
  const history = createMemoryHistory();

  if (pushUrl) {
    history.push(pushUrl);
  }

  store.dispatch = () => {};

  return (
    <redux.Provider store={store}>
      <Router history={history}>
        {children}
      </Router>
    </redux.Provider>
  );
};


