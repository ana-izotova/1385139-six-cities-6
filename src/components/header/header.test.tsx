import React from 'react';
import {render} from '@testing-library/react';
import Header from "./header";
import {TestMock} from "../../test-mocks/test-mock";

describe(`Header should render correctly`, () => {
  it(`Header should render correctly with unauthorized user`, () => {
    const {container} = render(
        <TestMock emptyStore={true}>
          <Header />
        </TestMock>
    );

    expect(container).toMatchSnapshot();
  });

  it(`Header should render correctly with authorized user`, () => {
    const {container} = render(
        <TestMock>
          <Header />
        </TestMock>
    );

    expect(container).toMatchSnapshot();
  });
});
