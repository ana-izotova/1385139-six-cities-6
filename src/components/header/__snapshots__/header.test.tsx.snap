// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render correctly Header should render correctly with authorized user 1`] = `
<div>
  <header
    class="header"
  >
    <div
      class="container"
    >
      <div
        class="header__wrapper"
      >
        <div
          class="header__left"
        >
          <a
            class="header__logo-link "
            href="/"
          >
            <img
              alt="6 cities logo"
              class="header__logo"
              src="img/logo.svg"
              style="width: 81px; height: 41px;"
            />
          </a>
        </div>
        <nav
          class="header__nav"
        >
          <ul
            class="header__nav-list"
          >
            <li
              class="header__nav-item user"
            >
              <a
                class="header__nav-link header__nav-link--profile"
                href="/favorites"
              >
                <div
                  class="header__avatar-wrapper user__avatar-wrapper"
                  style="background-image: url(avatar.jpg); border-radius: 50%;"
                />
                <span
                  class="header__user-name user__name"
                >
                  jondoe@mail.com
                </span>
              </a>
            </li>
            <li
              class="header__nav-item user"
            >
              <a
                class="header__nav-link header__nav-link--profile"
                href="/"
                style="margin-left: 10px;"
              >
                <span
                  class="header__user-name user__name"
                >
                  Выход
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
</div>
`;

exports[`Header should render correctly Header should render correctly with unauthorized user 1`] = `
<div>
  <header
    class="header"
  >
    <div
      class="container"
    >
      <div
        class="header__wrapper"
      >
        <div
          class="header__left"
        >
          <a
            class="header__logo-link "
            href="/"
          >
            <img
              alt="6 cities logo"
              class="header__logo"
              src="img/logo.svg"
              style="width: 81px; height: 41px;"
            />
          </a>
        </div>
        <nav
          class="header__nav"
        >
          <ul
            class="header__nav-list"
          >
            <li
              class="header__nav-item user"
            >
              <a
                class="header__nav-link header__nav-link--profile"
                href="/login"
              >
                <div
                  class="header__avatar-wrapper user__avatar-wrapper"
                />
                <span
                  class="header__user-name user__name"
                >
                  Sign in
                </span>
              </a>
            </li>
          </ul>
        </nav>
      </div>
    </div>
  </header>
</div>
`;
